File: cpp/src/core/trace.cpp
#include "navigation/core/trace.hpp"

#include <cmath>
#include <cstdio>

namespace navigation::core {
namespace {

// JSON numbers: emit integral values without a decimal point (so cells read as
// [row, col] ints) and others with enough precision to round-trip coordinates.
void write_num(std::ostream& os, double v) {
  if (std::isfinite(v) && std::abs(v) < 9e15 && v == std::floor(v)) {
    os << static_cast<long long>(v);
  } else {
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%.10g", v);
    os << buf;
  }
}

void write_str(std::ostream& os, const std::string& s) {
  os << '"';
  for (char c : s) {
    switch (c) {
      case '"': os << "\\\""; break;
      case '\\': os << "\\\\"; break;
      case '\n': os << "\\n"; break;
      case '\t': os << "\\t"; break;
      case '\r': os << "\\r"; break;
      default: os << c;
    }
  }
  os << '"';
}

void write_array(std::ostream& os, const std::vector<double>& a) {
  os << '[';
  for (size_t i = 0; i < a.size(); ++i) {
    if (i) os << ',';
    write_num(os, a[i]);
  }
  os << ']';
}

void write_data(std::ostream& os, const std::map<std::string, double>* data) {
  if (!data) return;  // absent (or empty) data omits the field entirely
  os << ",\"data\":{";
  bool first = true;
  for (const auto& [k, v] : *data) {
    if (!first) os << ',';
    first = false;
    write_str(os, k);
    os << ':';
    write_num(os, v);
  }
  os << '}';
}

}  // namespace

TraceRecorder::TraceRecorder(std::ostream& os) : os_(os), t0_(std::chrono::steady_clock::now()) {}

void TraceRecorder::begin_event(const char* event) {
  double t = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0_).count();
  os_ << "{\"seq\":" << seq_++ << ",\"t\":";
  write_num(os_, t);
  os_ << ",\"event\":\"" << event << '"';
}

void TraceRecorder::end_event() { os_ << "}\n"; }

void TraceRecorder::begin_planning_started(const std::string& algorithm,
                                           const std::string& map_path,
                                           const std::map<std::string, ParamValue>& params) {
  t0_ = std::chrono::steady_clock::now();
  begin_event("planning_started");
  os_ << ",\"algorithm\":";
  write_str(os_, algorithm);
  os_ << ",\"map\":";
  write_str(os_, map_path);
  os_ << ",\"params\":{";
  bool first = true;
  for (const auto& [k, v] : params) {
    if (!first) os_ << ',';
    first = false;
    write_str(os_, k);
    os_ << ':';
    std::visit(
        [&](const auto& val) {
          using T = std::decay_t<decltype(val)>;
          if constexpr (std::is_same_v<T, std::string>) {
            write_str(os_, val);
          } else if constexpr (std::is_same_v<T, bool>) {
            os_ << (val ? "true" : "false");
          } else {
            write_num(os_, static_cast<double>(val));
          }
        },
        v);
  }
  os_ << '}';
}

void TraceRecorder::planning_started(const std::string& algorithm, const std::string& map_path,
                                     const std::map<std::string, ParamValue>& params) {
  begin_planning_started(algorithm, map_path, params);
  end_event();
}

void TraceRecorder::planning_started(const std::string& algorithm, const std::string& map_path,
                                     const std::map<std::string, ParamValue>& params,
                                     const std::string& scenario) {
  begin_planning_started(algorithm, map_path, params);
  os_ << ",\"scenario\":";
  write_str(os_, scenario);
  end_event();
}

void TraceRecorder::planning_finished(bool success, const std::map<std::string, double>& metrics) {
  begin_event("planning_finished");
  os_ << ",\"success\":" << (success ? "true" : "false") << ",\"metrics\":{";
  bool first = true;
  for (const auto& [k, v] : metrics) {
    if (!first) os_ << ',';
    first = false;
    write_str(os_, k);
    os_ << ':';
    write_num(os_, v);
  }
  os_ << '}';
  end_event();
}

void TraceRecorder::ev_state(const char* event, const std::vector<double>& s, const double* cost,
                             const EventData* data) {
  begin_event(event);
  os_ << ",\"state\":";
  write_array(os_, s);
  if (cost) {
    os_ << ",\"cost\":";
    write_num(os_, *cost);
  }
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_rollout(const std::vector<double>& s, double cost, const EventData* data,
                               const std::vector<std::vector<double>>& rollout) {
  begin_event("candidate_evaluated");
  os_ << ",\"state\":";
  write_array(os_, s);
  os_ << ",\"cost\":";
  write_num(os_, cost);
  write_data(os_, data);
  os_ << ",\"rollout\":[";
  for (size_t i = 0; i < rollout.size(); ++i) {
    if (i) os_ << ',';
    write_array(os_, rollout[i]);
  }
  os_ << ']';
  end_event();
}

void TraceRecorder::ev_robot_moved(const std::vector<double>& s, int agent,
                                   const EventData* data) {
  begin_event("robot_moved");
  os_ << ",\"state\":";
  write_array(os_, s);
  os_ << ",\"agent\":" << agent;
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_velocity_obstacle(const std::vector<double>& s,
                                         const std::vector<std::vector<double>>& constraints,
                                         const EventData* data) {
  begin_event("velocity_obstacle");
  os_ << ",\"state\":";
  write_array(os_, s);
  if (!constraints.empty()) {
    os_ << ",\"constraints\":[";
    for (size_t i = 0; i < constraints.size(); ++i) {
      if (i) os_ << ',';
      write_array(os_, constraints[i]);
    }
    os_ << ']';
  }
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_edge(const char* event, const std::vector<double>& s,
                            const std::vector<double>& parent, const double* cost,
                            const EventData* data) {
  begin_event(event);
  os_ << ",\"state\":";
  write_array(os_, s);
  os_ << ",\"parent\":";
  write_array(os_, parent);
  if (cost) {
    os_ << ",\"cost\":";
    write_num(os_, *cost);
  }
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_bins(const char* event, const std::vector<double>& s,
                            const std::vector<double>& bins, const EventData* data) {
  begin_event(event);
  os_ << ",\"state\":";
  write_array(os_, s);
  os_ << ",\"bins\":";
  write_array(os_, bins);
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_band(const std::vector<std::vector<double>>& band, const EventData* data) {
  begin_event("band_updated");
  os_ << ",\"band\":[";
  for (size_t i = 0; i < band.size(); ++i) {
    if (i) os_ << ',';
    write_array(os_, band[i]);
  }
  os_ << ']';
  write_data(os_, data);
  end_event();
}

void TraceRecorder::ev_path(const std::vector<std::vector<double>>& path) {
  begin_event("path_found");
  os_ << ",\"path\":[";
  for (size_t i = 0; i < path.size(); ++i) {
    if (i) os_ << ',';
    write_array(os_, path[i]);
  }
  os_ << ']';
  end_event();
}

}  // namespace navigation::core

File: python/navigation/core/trace.py
"""Step-by-step trace recorder — the contract for visualization.

Mirrors the C++ `core/trace.hpp`. Emits JSON Lines per `spec/trace_schema.json`.
`seq` starts at 0 and increments per event; `t` is seconds since planning_started.
A null recorder must cost nothing on the hot path: callers guard every emit with
``if recorder is not None`` so a None recorder never runs any of this code.
"""

from __future__ import annotations

import json
import time
from collections.abc import Mapping, Sequence
from types import TracebackType
from typing import TextIO

from .params import ParamValue

# A serialized state is a numeric tuple -> JSON array (Cell ints, Point floats).
State = Sequence[float]
# Optional algorithm-specific extra info attached to an event (spec/trace_schema.json
# `data`). Numeric only, mirroring the C++ recorder's std::map<string, double>; viz may
# read it or ignore it. Never put info here that a common field already carries.
EventData = Mapping[str, float]


def _add_data(fields: dict[str, object], data: EventData | None) -> None:
    # Empty data carries no information, so it is elided to keep traces (and the
    # C++ recorder, which omits an empty map) byte-identical when unused.
    if data:
        fields["data"] = dict(data)


class TraceRecorder:
    def __init__(self, out: TextIO, owns: bool = False) -> None:
        self._out = out
        self._owns = owns
        self._seq = 0
        self._t0 = time.monotonic()

    def __enter__(self) -> TraceRecorder:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        self.close()

    def close(self) -> None:
        if self._owns:
            self._out.close()

    def _emit(self, event: str, fields: dict[str, object]) -> None:
        record: dict[str, object] = {"seq": self._seq, "t": round(time.monotonic() - self._t0, 6)}
        record["event"] = event
        record.update(fields)
        self._out.write(json.dumps(record, separators=(",", ":")) + "\n")
        self._seq += 1

    # --- events -----------------------------------------------------------
    def planning_started(
        self,
        algorithm: str,
        map_path: str,
        params: dict[str, ParamValue],
        scenario: str | None = None,
    ) -> None:
        fields: dict[str, object] = {"algorithm": algorithm, "map": map_path, "params": params}
        # scenario is local-demo only (replay recovers reference_path/goal from it); omit
        # rather than emit null so existing global-planning traces stay byte-identical.
        if scenario is not None:
            fields["scenario"] = scenario
        self._emit("planning_started", fields)

    def node_expanded(
        self, state: State, cost: float | None = None, data: EventData | None = None
    ) -> None:
        fields: dict[str, object] = {"state": list(state)}
        if cost is not None:
            fields["cost"] = cost
        _add_data(fields, data)
        self._emit("node_expanded", fields)

    def edge_added(
        self, state: State, parent: State, cost: float | None = None, data: EventData | None = None
    ) -> None:
        fields: dict[str, object] = {"state": list(state), "parent": list(parent)}
        if cost is not None:
            fields["cost"] = cost
        _add_data(fields, data)
        self._emit("edge_added", fields)

    def sample_drawn(self, state: State, data: EventData | None = None) -> None:
        fields: dict[str, object] = {"state": list(state)}
        _add_data(fields, data)
        self._emit("sample_drawn", fields)

    def rewire(self, state: State, parent: State, data: EventData | None = None) -> None:
        fields: dict[str, object] = {"state": list(state), "parent": list(parent)}
        _add_data(fields, data)
        self._emit("rewire", fields)

    def candidate_evaluated(
        self,
        state: State,
        cost: float,
        data: EventData | None = None,
        rollout: Sequence[State] | None = None,
    ) -> None:
        # rollout (spec/trace_schema.json): predicted trajectory polyline for
        # rollout-scoring planners (DWA). A top-level array like `bins` because
        # the numeric-only `data` map cannot carry arrays; omitted when absent
        # so existing traces stay byte-identical.
        fields: dict[str, object] = {"state": list(state), "cost": cost}
        _add_data(fields, data)
        if rollout is not None:
            fields["rollout"] = [list(p) for p in rollout]
        self._emit("candidate_evaluated", fields)

    def robot_moved(self, state: State, data: EventData | None = None) -> None:
        # Dynamic replanning (D* Lite): the robot's new executed cell. Also reused by
        # the local-planning simulator for every closed-loop tick's executed pose,
        # where `data` optionally carries the (v, omega) command that produced it.
        fields: dict[str, object] = {"state": list(state)}
        _add_data(fields, data)
        self._emit("robot_moved", fields)

    def force_computed(self, state: State, data: EventData | None = None) -> None:
        # Potential Fields (Khatib 1986): attractive/repulsive/resultant force
        # components at the current pose, one event per control tick.
        fields: dict[str, object] = {"state": list(state)}
        _add_data(fields, data)
        self._emit("force_computed", fields)

    def histogram_updated(
        self, state: State, bins: Sequence[float], data: EventData | None = None
    ) -> None:
        # VFH (Borenstein & Koren 1991): polar obstacle-density histogram. `bins` is
        # an array, which the numeric-only `data` map cannot carry, so it is a
        # dedicated top-level field.
        fields: dict[str, object] = {"state": list(state), "bins": list(bins)}
        _add_data(fields, data)
        self._emit("histogram_updated", fields)

    def band_updated(self, band: Sequence[Sequence[float]], data: EventData | None = None) -> None:
        # Elastic Bands / TEB (spec/trace_schema.json `band`): the deformed
        # bubble-chain or timed-pose-chain state, one event per control tick. No
        # `state` field -- unlike candidate_evaluated/histogram_updated, the band
        # itself *is* the event's subject, not an annotation on a robot pose.
        fields: dict[str, object] = {"band": [list(b) for b in band]}
        _add_data(fields, data)
        self._emit("band_updated", fields)

    def obstacle_revealed(self, state: State) -> None:
        # Dynamic replanning (D* Lite): a cell newly sensed as blocked.
        self._emit("obstacle_revealed", {"state": list(state)})

    def path_found(self, path: Sequence[State]) -> None:
        self._emit("path_found", {"path": [list(s) for s in path]})

    def planning_finished(self, success: bool, metrics: dict[str, float]) -> None:
        self._emit("planning_finished", {"success": success, "metrics": metrics})


def open_trace(path: str) -> TraceRecorder:
    """Open ``path`` for writing and return a recorder that owns the file."""
    return TraceRecorder(open(path, "w", encoding="utf-8"), owns=True)
